packages/ckeditor5-html-support/src/datafilter.ts

```diff
@@ -805,129 +807,133 @@ export interface DataFilterRegisterEvent {
-		consumedMatches.push( match );
+function normalizeGHSAttribute(
+	viewElement: ViewElement,
+	{ attributes, classes, styles }: {
+		attributes: Array<string>;
+		classes: Array<string>;
+		styles: Array<string>;
+	}
+): GHSViewAttributes | null {
+	if ( !attributes.length && !classes.length && !styles.length ) {
+		return null;
 	}
 
-	return consumedMatches;
-}
-
-/**
- * Removes attributes from the given match that were already consumed by other converters.
- */
-function removeConsumedAttributes( consumable: ViewConsumable, viewElement: ViewElement, match: MatchResult ) {
-	for ( const key of [ 'attributes', 'classes', 'styles' ] as const ) {
-		const attributes = match.match[ key ];
+	return {
+		...( attributes.length && {
+			attributes: normalizeAttributes( viewElement, attributes )
+		} ),
 
-		if ( !attributes ) {
-			continue;
-		}
+		...( styles.length && {
+			styles: normalizeStyles( viewElement, styles )
+		} ),
 
-		// Iterating over a copy of an array so removing items doesn't influence iteration.
-		for ( const value of Array.from( attributes ) ) {
-			if ( !consumable.test( viewElement, ( { [ key ]: [ value ] } ) ) ) {
-				removeItemFromArray( attributes, value );
-			}
-		}
-	}
+		...( classes.length && {
+			classes
+		} )
+	};
 }
 
 /**
- * Merges the result of {@link module:engine/view/matcher~Matcher#matchAll} method.
- *
- * @param matches
- * @returns Object with following properties:
- * - attributes Set with matched attribute names.
- * - styles Set with matched style names.
- * - classes Set with matched class names.
+ * Returns list of attributes as a normalized object.
  */
-function mergeMatchResults( matches: Array<MatchResult> ):
-{
-	attributes: Set<string>;
-	styles: Set<string>;
-	classes: Set<string>;
-} {
-	const matchResult = {
-		attributes: new Set<string>(),
-		classes: new Set<string>(),
-		styles: new Set<string>()
-	};
+function normalizeAttributes( viewElement: ViewElement, attributes: Iterable<string> ): Record<string, string> {
+	const attributesObject: Record<string, string> = {};
 
-	for ( const match of matches ) {
-		for ( const key in matchResult ) {
-			const values: Array<string> = match.match[ key as keyof typeof matchResult ] || [];
+	for ( const key of attributes ) {
+		const value = viewElement.getAttribute( key );
 
-			values.forEach( value => ( matchResult[ key as keyof typeof matchResult ] ).add( value ) );
+		if ( value !== undefined && isValidAttributeName( key ) ) {
+			attributesObject[ key ] = value;
 		}
 	}
 
-	return matchResult;
+	return attributesObject;
 }
 
 /**
- * Converts the given iterable object into an object.
+ * Returns list of styles as a normalized object (without redundant entries).
  */
-function iterableToObject( iterable: Set<string>, getValue: ( s: string ) => any ) {
-	const attributesObject: Record<string, any> = {};
+function normalizeStyles( viewElement: ViewElement, styles: Iterable<string> ): Record<string, string> {
```

Comment:
If the function returns a list we can add a prefix `get` to the name of function, WDYT?

So the name would be `getNormalizedStyles`.

Reply:
Changed to `getReducedStyles()`.